workers/unity/Packages/com.improbable.gdk.core/Dispatcher.cs

```diff
@@ -0,0 +1,494 @@
+    // callback removes itself, a call to InvokeAll iterates over the map to invoke each callback.
+    // This then indirectly invokes Remove, which makes the iteration crash.
+    //
+    // A simple solution to this problem would be to copy the map for each iteration, but this
+    // allocates a significant amount of memory each time. Instead, we fix this issue by postponing
+    // changes to the map until the end of the top-level call to this object. In the mentioned
+    // example, when InvokeAll indirectly calls Remove, the removal is not done immediately, but
+    // postponed to the end of the (top-level) InvokeAll call.
+    //
+    // The current callbacks are those in the map, plus toAdd, minus toRemove. The changes are
+    // merged via the UpdateGuard.
+    internal class Callbacks<T>
+    {
+        private SortedDictionary<ulong, Action<T>> map = new SortedDictionary<ulong, Action<T>>();
+
+        private SortedDictionary<ulong, Action<T>> mapReversed =
+            new SortedDictionary<ulong, Action<T>>(new DescendingComparer<ulong>());
+
+        private SortedDictionary<ulong, Action<T>> toAdd = new SortedDictionary<ulong, Action<T>>();
+        private SortedDictionary<ulong, uint> toRemove = new SortedDictionary<ulong, uint>();
+
+        // Current call level with 0 for the top-level call.
+        private int callDepth = 0;
+
+        public void Add(ulong key, Action<T> callback)
+        {
+            try
+            {
+                EnterUpdateGuard();
+                toAdd.Add(key, callback);
+                toRemove.Remove(key);
+            }
+            finally
+            {
+                ExitUpdateGuard();
+            }
+        }
+
+        public bool Remove(ulong key)
+        {
+            try
+            {
+                EnterUpdateGuard();
+
+                var isInMap = map.ContainsKey(key);
+                var hadPendingAdd = toAdd.Remove(key);
+
+                if (isInMap)
+                {
+                    var createdPendingRemove = !toRemove.ContainsKey(key);
+                    toRemove.Add(key, /* ignored */ 0);
+                    return createdPendingRemove;
+                }
+                else
+                {
+                    return hadPendingAdd;
+                }
+            }
+            finally
+            {
+                ExitUpdateGuard();
+            }
+        }
+
+        public void InvokeAll(T op)
+        {
+            try
+            {
+                EnterUpdateGuard();
+                foreach (var pair in map)
+                {
+                    pair.Value(op);
+                }
+            }
+            finally
+            {
+                ExitUpdateGuard();
+            }
+        }
+
+        public void InvokeAllReverse(T op)
+        {
+            try
+            {
+                EnterUpdateGuard();
+                foreach (var pair in map)
```

Comment:
We can save the performance things for later but this does need to be reversed.

Reply:
This should be `mapReversed`.

Reply:
Can fix 👍